File: packages/gbnf/src/rules-builder/parse-char.test.ts
import { parseChar } from "./parse-char";

describe('parseChar', () => {
  test.each([
    [
      'escaped 8-bit unicode char',
      'a',
      'a'.charCodeAt(0),
    ],
    [
      'escaped 8-bit unicode char',
      '9',
      '9'.charCodeAt(0),
    ],
  ])('parses char `%s`', (_key, char, codePoint) => {
    const grammar = `root ::= "${char}" "foo"`;
    expect(parseChar(grammar, `root ::= "`.length)).toEqual([codePoint, 1]);
  });

  test.each([
    [
      'escaped 8-bit unicode char',
      '\\x2A',
      '\x2A'.charCodeAt(0),
      4,
    ],
    [
      'escaped 16-bit unicode char',
      '\\u006F',
      '\u006F'.charCodeAt(0),
      6,
    ],
    [
      'escaped 32-bit unicode char',
      "\\U0001F4A9",
      128169,
      10,
    ],
    [
      'escaped tab char',
      "\\t",
      '\t'.charCodeAt(0),
      2,
    ],
    [
      'escaped new line char',
      "\\n",
      '\n'.charCodeAt(0),
      2,
    ],
    [
      'escaped \r char',
      "\\r",
      '\r'.charCodeAt(0),
      2,
    ],
    [
      'escaped quote char',
      "\\\"",
      '"'.charCodeAt(0),
      2,
    ],
    [
      'escaped [ char',
      "\\[",
      '['.charCodeAt(0),
      2,
    ],
    [
      'escaped ] char',
      "\\]",
      ']'.charCodeAt(0),
      2,
    ],
    [
      'escaped \\ char',
      "\\\\",
      '\\'.codePointAt(0),
      2,
    ],
  ])('parses %s `%s`', (key, escapedChar, codePoint, incPos) => {
    const grammar = `root ::= "${escapedChar}" "foo"`;
    expect(parseChar(grammar, `root ::= "`.length)).toEqual([codePoint, incPos]);
  });

  test('it throws on invalid input', () => {
    expect(() => parseChar('', 0)).toThrow();
    expect(() => parseChar('a', 1)).toThrow();
  });
});

File: packages/gbnfpy/gbnf/rules_builder/parse_char_test.py
import pytest

from .errors import GrammarParseError
from .parse_char import parse_char


@pytest.mark.parametrize(
    ("description", "char", "code_point"),
    [
        ("escaped 8-bit unicode char", "a", ord("a")),
        ("escaped 8-bit unicode char", "9", ord("9")),
    ],
)
def test_parse_char_simple(description, char, code_point):
    grammar = f'root ::= "{char}" "foo"'
    assert parse_char(grammar, len('root ::= "')) == (code_point, 1)


@pytest.mark.parametrize(
    ("description", "escaped_char", "code_point", "inc_pos"),
    [
        ("escaped 8-bit unicode char", "\\x2A", ord("\x2A"), 4),
        ("escaped 16-bit unicode char", "\\u006F", ord("\u006F"), 6),
        ("escaped 32-bit unicode char", "\\U0001F4A9", 128169, 10),
        ("escaped tab char", "\\t", ord("\t"), 2),
        ("escaped new line char", "\\n", ord("\n"), 2),
        ("escaped \r char", "\\r", ord("\r"), 2),
        ("escaped quote char", '\\"', ord('"'), 2),
        ("escaped [ char", "\\[", ord("["), 2),
        ("escaped ] char", "\\]", ord("]"), 2),
        ("escaped \\ char", "\\\\", ord("\\"), 2),
    ],
)
def test_parse_char_complex(description, escaped_char, code_point, inc_pos):
    grammar = f'root ::= "{escaped_char}" "foo"'
    assert parse_char(grammar, len('root ::= "')) == (code_point, inc_pos)


def test_parse_char_throws():
    with pytest.raises(GrammarParseError):
        parse_char("", 0)
    with pytest.raises(GrammarParseError):
        parse_char("a", 1)
